cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -248,11 +248,11 @@ func Run(c *config.CompletedConfig, stopCh <-chan struct{}) error {
 			for i, rpKubeConfig := range c.ResourceProviderKubeconfigs {
 				rpId := "rp" + strconv.Itoa(i)
 
-				clientConfigs := restclient.AnonymousClientConfig(rpKubeConfig)
+				clientConfigs := *rpKubeConfig
```

Comment:
This is shallow copy. Please use the CopyConfig function so that we don't unexpectedly update userAgent for other client that is using rpKubeConfig

Reply:
My test yield the following:

@@ -241,7 +241,7 @@ func Run(c *config.CompletedConfig, stopCh <-chan struct{}) error {

                // start client to resource providers
                userAgent := getUserAgent(controllerContext)
-               klog.Infof("=== user agent [%v]", userAgent) // TODO: remove before merge into master
+               klog.Infof("==== setting user agent [%v]", userAgent) // TODO: remove before merge into master
                if len(c.ResourceProviderKubeconfigs) > 0 {
                        controllerContext.ResourceProviderClients = make(map[string]clientset.Interface)
                        controllerContext.ResourceProviderNodeInformers = make(map[string]coreinformers.NodeInformer)
@@ -249,9 +249,14 @@ func Run(c *config.CompletedConfig, stopCh <-chan struct{}) error {
                                rpId := "rp" + strconv.Itoa(i)

                                clientConfigs := *rpKubeConfig
+                               klog.Infof("==== user agent 1 for kubeconfig [%v]", rpKubeConfig.GetConfig().UserAgent)
+                               klog.Infof("==== user agent 1 for clientConfigs [%v]", clientConfigs.GetConfig().UserAgent)
+
                                for _, config := range clientConfigs.GetAllConfigs() {
                                        config.UserAgent = userAgent
                                }
+                               klog.Infof("==== user agent 2 for kubeconfig [%v]", rpKubeConfig.GetConfig().UserAgent)
+                               klog.Infof("==== user agent 2 for clientConfigs [%v]", clientConfigs.GetConfig().UserAgent)
                                rpClient := clientset.NewForConfigOrDie(&clientConfigs)

                                resourceInformerFactory := informers.NewSharedInformerFactory(rpClient, 0)
ubuntu@ip-172-30-0-148:~/go/src/dev-arktos$ cat /tmp/kube-controller-manager.log | grep ====
I0730 21:16:50.456832    4860 controllermanager.go:244] ==== setting user agent [kcm-tenant-provider]
I0730 21:16:50.456842    4860 controllermanager.go:252] ==== user agent 1 for kubeconfig []
I0730 21:16:50.456849    4860 controllermanager.go:253] ==== user agent 1 for clientConfigs []
I0730 21:16:50.456856    4860 controllermanager.go:258] ==== user agent 2 for kubeconfig [kcm-tenant-provider]
I0730 21:16:50.456863    4860 controllermanager.go:259] ==== user agent 2 for clientConfigs [kcm-tenant-provider]
I0730 21:16:50.457363    4860 controllermanager.go:252] ==== user agent 1 for kubeconfig []
I0730 21:16:50.457370    4860 controllermanager.go:253] ==== user agent 1 for clientConfigs []
I0730 21:16:50.457377    4860 controllermanager.go:258] ==== user agent 2 for kubeconfig [kcm-tenant-provider]
I0730 21:16:50.457383    4860 controllermanager.go:259] ==== user agent 2 for clientConfigs [kcm-tenant-provider]

Reply:
Also, please note previous changes that update user agent "node-controller" to "node-ipam-controller" had permission issue. I think our newly introduced user agent "kcm-tenant-provider" and "kcm-resource-provider" would also have some permission issue. Might need to assign permission in plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go

Reply:
Line 251: clientConfigs := restclient.CopyConfigs(rpKubeConfig)

Line 255: rpClient := clientset.NewForConfigOrDie(clientConfigs)

Reply:
uses CopyConfigs to deep-copy config, as suggested by peer feedback.